File: typescript/src/agent/mcp/provider.ts
// =============================================================================
// MCP Tool Provider — connects to MCP servers and exposes tools to the agent
// =============================================================================

import type { AgentTool, ToolProvider } from "../types.js";
import { MCPClient } from "./client.js";
import type { MCPServerConfig } from "./types.js";

/** Normalize a name for use in tool identifiers (replace non-alphanumeric with underscore). */
function normalizeName(name: string): string {
  return name.replace(/[^a-zA-Z0-9_]/g, "_");
}

/**
 * Sanitize a JSON Schema so it is accepted by the OpenResponses API.
 *
 * Some MCP servers (e.g. Composio) emit `type: ["string", "null"]` which is
 * valid JSON Schema but rejected by many LLM APIs that expect a single string.
 * This recursively converts array types to their first non-null entry.
 */
function sanitizeSchema(schema: Record<string, unknown>): Record<string, unknown> {
  const result = { ...schema };

  if (Array.isArray(result.type)) {
    const nonNull = result.type.filter((t: string) => t !== "null");
    result.type = nonNull.length === 1 ? nonNull[0] : nonNull.length > 0 ? nonNull[0] : "string";
  }

  if (result.properties && typeof result.properties === "object") {
    const props: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(result.properties as Record<string, unknown>)) {
      props[key] =
        value && typeof value === "object"
          ? sanitizeSchema(value as Record<string, unknown>)
          : value;
    }
    result.properties = props;
  }

  if (result.items && typeof result.items === "object") {
    result.items = sanitizeSchema(result.items as Record<string, unknown>);
  }

  return result;
}

/**
 * MCP tool provider — connects to one MCP server, discovers tools, and
 * wraps them as `AgentTool` instances for the agent loop.
 *
 * Tool naming follows the `mcp__serverName__toolName` convention.
 */
export class MCPToolProvider implements ToolProvider {
  readonly type = "tool_provider" as const;

  private readonly config: MCPServerConfig;
  private client: MCPClient | null = null;

  constructor(config: MCPServerConfig) {
    this.config = config;
  }

  async setup(): Promise<AgentTool[]> {
    this.client = new MCPClient(this.config);
    await this.client.connect();

    const mcpTools = await this.client.listTools();
    return mcpTools.map((t) => this.wrapTool(t));
  }

  async teardown(): Promise<void> {
    if (this.client) {
      await this.client.disconnect();
      this.client = null;
    }
  }

  private wrapTool(mcpTool: {
    name: string;
    description: string;
    inputSchema: Record<string, unknown>;
  }): AgentTool {
    const serverPrefix = normalizeName(this.config.name);
    const toolSuffix = normalizeName(mcpTool.name);
    const qualifiedName = `mcp__${serverPrefix}__${toolSuffix}`;
    const client = this.client!;
    const originalName = mcpTool.name;

    return {
      name: qualifiedName,
      description: mcpTool.description,
      parameters: sanitizeSchema(mcpTool.inputSchema),
      execute: async (input: unknown) => {
        const args =
          typeof input === "object" && input !== null
            ? (input as Record<string, unknown>)
            : input === undefined
              ? {}
              : { value: input };
        return client.callTool(originalName, args);
      },
    };
  }
}

File: python/src/opperai/agent/mcp/_provider.py
"""MCP Tool Provider — connects to MCP servers and exposes tools to the agent."""

from __future__ import annotations

import re
from typing import Any

from .._types import AgentTool
from ._client import MCPClient
from ._types import MCPServerConfig


def _normalize_name(name: str) -> str:
    """Normalize a name for use in tool identifiers (replace non-alphanumeric with underscore)."""
    return re.sub(r"[^a-zA-Z0-9_]", "_", name)


def _sanitize_schema(schema: dict[str, Any]) -> dict[str, Any]:
    """Sanitize a JSON Schema so it is accepted by the OpenResponses API.

    Some MCP servers emit ``type: ["string", "null"]`` which is valid JSON Schema
    but rejected by many LLM APIs that expect a single string. This recursively
    converts array types to their first non-null entry.
    """
    result = dict(schema)

    if isinstance(result.get("type"), list):
        non_null = [t for t in result["type"] if t != "null"]
        result["type"] = non_null[0] if len(non_null) == 1 else (non_null[0] if non_null else "string")

    if isinstance(result.get("properties"), dict):
        result["properties"] = {
            key: _sanitize_schema(value) if isinstance(value, dict) else value
            for key, value in result["properties"].items()
        }

    if isinstance(result.get("items"), dict):
        result["items"] = _sanitize_schema(result["items"])

    return result


class MCPToolProvider:
    """MCP tool provider — connects to one MCP server, discovers tools, and
    wraps them as ``AgentTool`` instances for the agent loop.

    Tool naming follows the ``mcp__<serverName>__<toolName>`` convention.
    """

    def __init__(self, config: MCPServerConfig) -> None:
        self._config = config
        self._client: MCPClient | None = None

    async def setup(self) -> list[AgentTool]:
        client = MCPClient(self._config)
        self._client = client
        await client.connect()

        mcp_tools = await client.list_tools()
        return [self._wrap_tool(t) for t in mcp_tools]

    async def teardown(self) -> None:
        if self._client is not None:
            await self._client.disconnect()
            self._client = None

    def _wrap_tool(self, mcp_tool: Any) -> AgentTool:
        server_prefix = _normalize_name(self._config.name)
        tool_suffix = _normalize_name(mcp_tool.name)
        qualified_name = f"mcp__{server_prefix}__{tool_suffix}"
        client = self._client
        original_name = mcp_tool.name

        async def execute(**kwargs: Any) -> Any:
            assert client is not None
            return await client.call_tool(original_name, kwargs if kwargs else None)

        return AgentTool(
            name=qualified_name,
            description=mcp_tool.description,
            parameters=_sanitize_schema(mcp_tool.input_schema),
            execute=execute,
        )
